keps/sig-node/20191127-cri-containerstats-enhancements.md

```diff
@@ -0,0 +1,201 @@
+authors:
+  - "@fuweid"
+owning-sig: sig-node
+participating-sigs:
+  - sig-node
+  - sig-instrumentation
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-27
+last-updated: 2019-12-09
+status: provisional
+---
+
+# Enhance ContainerStats message in CRI-API
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Implementation Details/Notes/Constraints](#implementation-detailsnotesconstraints)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+<!-- /toc -->
+
+## Summary
+
+This enhancement is to extend `CpuUsage` and `MemoryUsage` fields in
+`ContainerStats`. And also add new `ProcessUsage`/`NetworkUsage` field in
+`ContainerStats`. The Container Runtime Engine will cover more metrics about
+managed container and let the metric collectors focus on node-level metrics or
+monitoring metrics.
+
+## Motivation
+
+Since more and more container runtimes shows up, like `kata-container`/
+`gVisor`/ `Firecracker`/ `Windows Containers`, the way to collect metrics from
+these containers is quite different from runC. For the third-party metric
+collectors, they have to maintain the knowledge about how to collect metrics
+from each kind of container runtimes. And more, kubelet codebase focus on
+core-metrics export (defined in [Kubernetes monitoring architecture#Terminology](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#terminology)),
+and there is no powerful standard interface for container level metrics.
+It will bring burden to end-users who wants to build powerful monitor pipeline
+(defined in [Kubernetes monitoring architecture#Executive Summary](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) for different container runtimes.
+
+Based on this case, we can extend Container Runtime Interface(CRI) to define
+the container level metrics. Each container runtime engine, which implemented
+CRI, knows lifecycle and metrics about each managed container better than any
+external metric collector. For the third-party metric collectors, they just
+need to follow the `ContainerStats` API defined by CRI. With standard CRI,
+it is easy for the collector from monitor pipeline (defined in [Kubernetes monitoring architecture](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/instrumentation/monitoring_architecture.md#executive-summary)) to collect metrics from different kind of runtimes.
+
+### Goals
+
+- Enhance Container Runtime Interface(CRI) to provide more metrics about
```

Comment:
Would be good to define what metrics we want to expose, with some arguments on why. e.g. why CRI should not expose GPU metrics. How runtime custom metrics should be exposed?

Reply:
We did some work to enable out-of-tree device monitoring back in 1.13 with https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/compute-device-assignment.md.  NVidia has provided a [daemonset](https://github.com/NVIDIA/gpu-monitoring-tools/blob/master/exporters/prometheus-dcgm/k8s/node-exporter/pod-gpu-node-exporter-daemonset.yaml) that uses this API to provide GPU metrics.  It has a superset of the metrics included in the summary API and cAdvisor.

Reply:
Interesting, thanks for link. With this question I wanted to start discussion on what should be part of CRI and what should be provided by daemonsets that you linked.

I would be interesting to have some higher level view on how kubernetes plug-able infrastructure provides metrics. It could be beneficial if we can decompose metrics by interface responsible for them. e.g. network metrics by CNI, storage by CSI. That would remove problem of putting every possible metric that someone needs in CRI.

Reply:
To my knowledge CSI already has some of this not sure in what state it is though and how enforced it is.

Reply:
> It could be beneficial if we can decompose metrics by interface responsible for them. e.g. network metrics by CNI, storage by CSI. That would remove problem of putting every possible metric that someone needs in CRI.

Make senses to me. Will update it.